UnicodeEncodeError: 'latin-1' codec can't encode character

The issue came up again, this time pytest reveals the content of the string that caused the problem (note the backticks around 'domount')

```
s = 'chown: changing ownership of ‘domount’: Operation not permitted\n'

    @staticmethod
    def b(s):
>       return s.encode("latin-1")
E       UnicodeEncodeError: 'latin-1' codec can't encode character '\u2018' in position 29: ordinal not in range(256)

../../plumbum/plumbum/lib.py:37: UnicodeEncodeError
```

--- a/plumbum/commands/base.py
+++ b/plumbum/commands/base.py
@@ -214,44 +214,6 @@ class BaseCommand(object):
         with self.bgrun(args, **kwargs) as p:
             return p.run()
 
-    def iter_lines(self, args = (), **kwargs):
-        """Runs the given command (equivalent to run()) and yields a tuples of (out, err) line pairs.
-        If the exit code of the process does not match the expected one, :class:`ProcessExecutionError
-        <plumbum.commands.ProcessExecutionError>` is raised.
-
-        :param args: Any arguments to be passed to the process (a tuple)
-
-        :param retcode: The expected return code of this process (defaults to 0).
-                        In order to disable exit-code validation, pass ``None``. It may also
-                        be a tuple (or any iterable) of expected exit codes.
-
-                        .. note:: this argument must be passed as a keyword argument.
-
-        :param timeout: The maximal amount of time (in seconds) to allow the process to run.
-                        ``None`` means no timeout is imposed; otherwise, if the process hasn't
-                        terminated after that many seconds, the process will be forcefully
-                        terminated an exception will be raised
-
-                        .. note:: this argument must be passed as a keyword argument.
-
-        :param linesize: Maximum number of characters to read from stdout/stderr at each iteration.
-                        ``-1`` (default) reads until a b'\\n' is encountered.
-
-                        .. note:: this argument must be passed as a keyword argument.
-
-        :param kwargs: Any keyword-arguments to be passed to the ``Popen`` constructor
-
-        :returns: An iterator of (out, err) line tuples.
-        """
-
-        retcode = kwargs.pop("retcode", 0)
-        timeout = kwargs.pop("timeout", None)
-        linesize = kwargs.pop("linesize", -1)
-
-        with self.bgrun(args, **kwargs) as popen:
-            for lines in iter_lines(popen, retcode=retcode, timeout=timeout, linesize=linesize):
-                yield lines
-
 
 class BoundCommand(BaseCommand):
     __slots__ = ["cmd", "args"]

--- a/plumbum/commands/processes.py
+++ b/plumbum/commands/processes.py
@@ -5,6 +5,7 @@ import heapq
 from subprocess import Popen
 from threading import Thread
 from plumbum.lib import IS_WIN32, six
+
 try:
     from queue import Queue, Empty as QueueEmpty
 except ImportError:
@@ -51,6 +52,55 @@ if not hasattr(Popen, "kill"):
         Popen.kill = _Popen_kill
         Popen.terminate = _Popen_kill
         Popen.send_signal = _Popen_send_signal
+
+#===================================================================================================
+# utility functions
+#===================================================================================================
+def _check_process(proc, retcode, timeout, stdout, stderr):
+    if getattr(proc, "_timed_out", False):
+        raise ProcessTimedOut("Process did not terminate within %s seconds" % (timeout,),
+            getattr(proc, "argv", None))
+
+    if retcode is not None:
+        if hasattr(retcode, "__contains__"):
+            if proc.returncode not in retcode:
+                raise ProcessExecutionError(getattr(proc, "argv", None), proc.returncode,
+                    stdout, stderr)
+        elif proc.returncode != retcode:
+            raise ProcessExecutionError(getattr(proc, "argv", None), proc.returncode,
+                stdout, stderr)
+    return proc.returncode, stdout, stderr
+
+def _iter_lines(proc, decode, linesize):
+    try:
+        from selectors import DefaultSelector, EVENT_READ
+    except ImportError:
+        # Pre Python 3.4 implementation
+        from select import select
+        def selector():
+            while True:
+                rlist, _, _ = select([proc.stdout, proc.stderr], [], [])
+                for stream in rlist:
+                    yield (stream is proc.stderr), decode(stream.readline(linesize))
+    else:
+        # Python 3.4 implementation
+        def selector():
+            sel = DefaultSelector()
+            sel.register(proc.stdout, EVENT_READ, 0)
+            sel.register(proc.stderr, EVENT_READ, 1)
+            while True:
+                for key, mask in sel.select():
+                    yield key.data, decode(key.fileobj.readline(linesize))
+
+    for ret in selector():
+        yield ret
+        if proc.poll() is not None:
+            break
+    for line in proc.stdout:
+        yield 0, decode(line)
+    for line in proc.stderr:
+        yield 1, decode(line)
+
 
 #===================================================================================================
 # Exceptions
@@ -207,7 +257,7 @@ def run_proc(proc, retcode, timeout = None):
 #===================================================================================================
 # iter_lines
 #===================================================================================================
-def iter_lines(proc, retcode = 0, timeout = None, linesize = -1):
+def iter_lines(proc, retcode = 0, timeout = None, linesize = -1, _iter_lines = _iter_lines):
     """Runs the given process (equivalent to run_proc()) and yields a tuples of (out, err) line pairs.
     If the exit code of the process does not match the expected one, :class:`ProcessExecutionError
     <plumbum.commands.ProcessExecutionError>` is raised.
@@ -229,39 +279,14 @@ def iter_lines(proc, retcode = 0, timeout = None, linesize = -1):
 
     encoding = getattr(proc, "encoding", None)
     if encoding:
-        read_stream = lambda s: s.readline(linesize).decode(encoding).rstrip()
+        decode = lambda s: s.decode(encoding).rstrip()
     else:
-        read_stream = lambda s: s.readline(linesize)
+        decode = lambda s: s
 
     _register_proc_timeout(proc, timeout)
 
-    try:
-        from selectors import DefaultSelector, EVENT_READ
-    except ImportError:
-        # Pre Python 3.4 implementation
-        def _iter_lines():
-            from select import select
-            while True:
-                rlist, _, _ = select([proc.stdout, proc.stderr], [], [])
-                for stream in rlist:
-                    yield (stream is proc.stderr), read_stream(stream)
-                if proc.poll() is not None:
-                    break
-    else:
-        # Python 3.4 implementation
-        sel = DefaultSelector()
-
-        sel.register(proc.stdout, EVENT_READ, 0)
-        sel.register(proc.stderr, EVENT_READ, 1)
-        def _iter_lines():
-            while True:
-                for key, mask in sel.select():
-                    yield key.data, read_stream(key.fileobj)
-                if proc.poll() is not None:
-                    break
-
     buffers = [StringIO(), StringIO()]
-    for t, line in _iter_lines():
+    for t, line in _iter_lines(proc, decode, linesize):
         ret = [None, None]
         ret[t] = line
         buffers[t].write(line + "\n")
@@ -269,22 +294,3 @@ def iter_lines(proc, retcode = 0, timeout = None, linesize = -1):
 
     # this will take care of checking return code and timeouts
     _check_process(proc, retcode, timeout, *(s.getvalue() for s in buffers))
-
-
-#===================================================================================================
-# _check_process
-#===================================================================================================
-def _check_process(proc, retcode, timeout, stdout, stderr):
-    if getattr(proc, "_timed_out", False):
-        raise ProcessTimedOut("Process did not terminate within %s seconds" % (timeout,),
-            getattr(proc, "argv", None))
-
-    if retcode is not None:
-        if hasattr(retcode, "__contains__"):
-            if proc.returncode not in retcode:
-                raise ProcessExecutionError(getattr(proc, "argv", None), proc.returncode,
-                    stdout, stderr)
-        elif proc.returncode != retcode:
-            raise ProcessExecutionError(getattr(proc, "argv", None), proc.returncode,
-                stdout, stderr)
-    return proc.returncode, stdout, stderr

--- a/plumbum/lib.py
+++ b/plumbum/lib.py
@@ -34,7 +34,7 @@ class six(object):
         
         @staticmethod
         def b(s):
-            return s.encode("latin-1")
+            return s.encode("latin-1", "replace")
         @staticmethod
         def u(s):
             return s

--- a/plumbum/machines/local.py
+++ b/plumbum/machines/local.py
@@ -6,6 +6,7 @@ import logging
 import time
 import platform
 import re
+from functools import partial
 from plumbum.path.local import LocalPath, LocalWorkdir
 from tempfile import mkdtemp
 from contextlib import contextmanager
@@ -14,6 +15,7 @@ from plumbum.commands import CommandNotFound, ConcreteCommand
 from plumbum.machines.session import ShellSession
 from plumbum.lib import ProcInfo, IS_WIN32, six
 from plumbum.commands.daemons import win32_daemonize, posix_daemonize
+from plumbum.commands.processes import iter_lines
 from plumbum.machines.env import BaseEnv
 
 if sys.version_info >= (3, 2):
@@ -29,6 +31,10 @@ else:
         from subprocess import Popen, PIPE
         has_new_subprocess = False
 
+class IterablePopen(Popen):
+    iter_lines = iter_lines
+    def __iter__(self):
+        return self.iter_lines()
 
 logger = logging.getLogger("plumbum.local")
 
@@ -255,7 +261,7 @@ class LocalMachine(object):
             argv, executable = self._as_user_stack[-1](argv)
 
         logger.debug("Running %r", argv)
-        proc = Popen(argv, executable = str(executable), stdin = stdin, stdout = stdout,
+        proc = IterablePopen(argv, executable = str(executable), stdin = stdin, stdout = stdout,
             stderr = stderr, cwd = str(cwd), env = env, **kwargs)  # bufsize = 4096
         proc._start_time = time.time()
         proc.encoding = self.encoding

--- a/plumbum/machines/paramiko_machine.py
+++ b/plumbum/machines/paramiko_machine.py
@@ -6,6 +6,7 @@ from plumbum.machines.session import ShellSession
 from plumbum.lib import _setdoc, six
 from plumbum.path.local import LocalPath
 from plumbum.path.remote import RemotePath
+from plumbum.commands.processes import iter_lines
 try:
     # Sigh... we need to gracefully-import paramiko for Sphinx builds, etc
     import paramiko
@@ -95,6 +96,12 @@ class ParamikoPopen(object):
         stderr = six.b("").join(six.b(s) for s in stderr)
         return stdout, stderr
 
+    def iter_lines(self, timeout=None, **kwargs):
+        if timeout is not None:
+            raise NotImplementedError("The 'timeout' parameter is not supported with ParamikoMachine")
+        return iter_lines(self, _iter_lines=_iter_lines, **kwargs)
+
+    __iter__ = iter_lines
 
 class ParamikoMachine(BaseRemoteMachine):
     """
@@ -149,7 +156,6 @@ class ParamikoMachine(BaseRemoteMachine):
             raise NotImplementedError("Not supported with ParamikoMachine")
         def __lshift__(self, *_):
             raise NotImplementedError("Not supported with ParamikoMachine")
-
 
     def __init__(self, host, user = None, port = None, password = None, keyfile = None,
             load_system_host_keys = True, missing_host_policy = None, encoding = "utf8",
@@ -223,7 +229,7 @@ class ParamikoMachine(BaseRemoteMachine):
         argv.extend(args.formulate())
         cmdline = " ".join(argv)
         logger.debug(cmdline)
-        si, so, se = self._client.exec_command(cmdline, 1)
+        si, so, se = streams = self._client.exec_command(cmdline, 1)
         return ParamikoPopen(argv, si, so, se, self.encoding, stdin_file = stdin,
             stdout_file = stdout, stderr_file = stderr)
 
@@ -327,4 +333,39 @@ class SocketCompatibleChannel(object):
         return self._chan.recv(count)
 
 
+###################################################################################################
+# Custom iter_lines for paramiko.Channel
+###################################################################################################
+def _iter_lines(proc, decode, linesize):
 
+    try:
+        from selectors import DefaultSelector, EVENT_READ
+    except ImportError:
+        # Pre Python 3.4 implementation
+        from select import select
+        def selector():
+            while True:
+                rlist, _, _ = select([proc.stdout.channel], [], [])
+                for _ in rlist:
+                    yield
+    else:
+        # Python 3.4 implementation
+        def selector():
+            sel = DefaultSelector()
+            sel.register(proc.stdout.channel, EVENT_READ)
+            while True:
+                for key, mask in sel.select():
+                    yield
+
+    for _ in selector():
+        if proc.stdout.channel.recv_ready():
+            yield 0, decode(six.b(proc.stdout.readline(linesize)))
+        if proc.stdout.channel.recv_stderr_ready():
+            yield 1, decode(six.b(proc.stderr.readline(linesize)))
+        if proc.poll() is not None:
+            break
+
+    for line in proc.stdout:
+        yield 0, decode(six.b(line))
+    for line in proc.stderr:
+        yield 1, decode(six.b(line))

--- a/plumbum/machines/paramiko_machine.py
+++ b/plumbum/machines/paramiko_machine.py
@@ -218,11 +218,11 @@ class ParamikoMachine(BaseRemoteMachine):
         return ShellSession(proc, self.encoding, isatty)
 
     @_setdoc(BaseRemoteMachine)
-    def popen(self, args, stdin = None, stdout = None, stderr = None, new_session = False):
+    def popen(self, args, stdin = None, stdout = None, stderr = None, new_session = False, cwd = None):
         # new_session is ignored for ParamikoMachine
         argv = []
         envdelta = self.env.getdelta()
-        argv.extend(["cd", str(self.cwd), "&&"])
+        argv.extend(["cd", str(cwd or self.cwd), "&&"])
         if envdelta:
             argv.append("env")
             argv.extend("%s=%s" % (k, v) for k, v in envdelta.items())

--- a/tests/test_local.py
+++ b/tests/test_local.py
@@ -199,7 +199,7 @@ class LocalMachineTest(unittest.TestCase):
         from plumbum.cmd import ping
 
         try:
-            for i, (out, err) in enumerate(ping["127.0.0.1", "-i", 0.5].iter_lines(timeout=2)):
+            for i, (out, err) in enumerate(ping["127.0.0.1", "-i", 0.5].popen().iter_lines(timeout=2)):
                 print("out:", out)
                 print("err:", err)
         except ProcessTimedOut:
@@ -211,7 +211,7 @@ class LocalMachineTest(unittest.TestCase):
     def test_iter_lines_error(self):
         from plumbum.cmd import ls
         try:
-            for i, lines in enumerate(ls["--bla"].iter_lines()):
+            for i, lines in enumerate(ls["--bla"].popen()):
                 pass
             self.assertEqual(i, 1)
         except ProcessExecutionError:

--- a/tests/test_remote.py
+++ b/tests/test_remote.py
@@ -1,10 +1,11 @@
 from __future__ import with_statement
+import sys
 import os
 import socket
 import unittest
 import time
 import logging
-from plumbum import RemotePath, SshMachine, ProcessExecutionError, local
+from plumbum import RemotePath, SshMachine, ProcessExecutionError, local, ProcessTimedOut
 from plumbum import CommandNotFound
 from plumbum.lib import six
 
@@ -135,6 +136,35 @@ s.close()
     def test_contains(self):
         with self._connect() as rem:
             self.assertTrue("ls" in rem, "Expected to find `ls`")
+
+    def test_iter_lines_timeout(self):
+        with self._connect() as rem:
+            try:
+                for i, (out, err) in enumerate(rem["ping"]["127.0.0.1", "-i", 0.5].popen().iter_lines(timeout=2)):
+                    print("out:", out)
+                    print("err:", err)
+            except NotImplementedError:
+                try:
+                    self.skipTest(sys.exc_info()[1])
+                except AttributeError:
+                    return
+            except ProcessTimedOut:
+                self.assertTrue(i > 3)
+            else:
+                self.fail("Expected a timeout")
+
+
+    def test_iter_lines_error(self):
+        with self._connect() as rem:
+            try:
+                for i, lines in enumerate(rem["ls"]["--bla"].popen()):
+                    pass
+                self.assertEqual(i, 1)
+            except ProcessExecutionError:
+                ex = sys.exc_info()[1]
+                self.assertTrue(ex.stderr.startswith("/bin/ls: unrecognized option '--bla'"))
+            else:
+                self.fail("Expected an execution error")
 
 
 class RemoteMachineTest(unittest.TestCase, BaseRemoteMachineTest):

